apps/05Mastermind.js

```diff
@@ -4,75 +4,106 @@ var assert = require('assert');
 var colors = require('colors/safe');
 var readline = require('readline');
 var rl = readline.createInterface({
-  input: process.stdin,
-  output: process.stdout
+    input: process.stdin,
+    output: process.stdout
 });
 
 var board = [];
 var solution = '';
 var letters = ['a', 'b', 'c', 'd', 'e', 'f', 'g', 'h'];
 
+var move = 0;
+
 function printBoard() {
-  for (var i = 0; i < board.length; i++) {
-    console.log(board[i]);
-  }
+    for (var i = 0; i < board.length; i++) {
+        console.log(board[i]);
+    }
 }
 
 function generateSolution() {
-  for (var i = 0; i < 4; i++) {
-    var randomIndex = getRandomInt(0, letters.length);
-    solution += letters[randomIndex];
-  }
+    for (var i = 0; i < 4; i++) {
+        var randomIndex = getRandomInt(0, letters.length);
+        solution += letters[randomIndex];
+    }
+
 }
 
 function getRandomInt(min, max) {
-  return Math.floor(Math.random() * (max - min)) + min;
+    return Math.floor(Math.random() * (max - min)) + min;
 }
 
-function generateHint() {
-  // your code here
+function generateHint(solution, guess) {
+    var solutionArray = solution.split("");
+    var guessArray = guess.split("");
+    var correctLetterLocations = 0;
+    var correctLetters = 0;
+
+    for (var i = 0; i < solutionArray.length; i++) {
+        if (solutionArray[i] === guessArray[i]) {
+            correctLetterLocations++;
+            solutionArray[i] = null;
+        }
+    }
+
+    for (var i = 0; i < solutionArray.length; i++) {
+        var targetIndex = solutionArray.indexOf(guessArray[i]);
+        if (targetIndex > -1) {
+            correctLetters++;
+            solutionArray[targetIndex] = null;
+        }
+    }
+    return correctLetterLocations + "-" + correctLetters;
 }
 
 function mastermind(guess) {
-  // your code here
+    if (guess === solution) {
+        return 'You guessed it!';
+    }
+    var hint = generateHint(solution, guess);
+    board.push(hint + "" + guess);
+
+    if (board.length === 10) {
+        return "You ran out of turns. The solution was " + solution;
+    }
+    return guess;
```

Comment:
Noooooiccceee. You're killing it!

![nicejob](http://reactiongif.org/wp-content/uploads/GIF/2014/08/GIF--Clap-applause-good-job-nice-one-clapping-laugh-Mike-Tyson-GIF.gif)